ledger/README.md

```diff
@@ -7,3 +7,59 @@ This package provides two ledger implementations:
+### binary Merkle tree
+
+![binary Merkle tree image](/ledger/docs/binary_merkle_tree.png?raw=true "binary Merkle tree" )
+
+In this context a *binary Merkle tree* is defined as [perfect binary tree](https://xlinux.nist.gov/dads/HTML/perfectBinaryTree.html) with a specific height including three type of nodes:
+
+- leaf nodes: holds a payload (data), a path (where the node is located), and a hash value (hash of path and payload content)
+
+- empty leaf nodes: doesn't hold any data and only stores a path, and a default hash value based on the height of tree
+
+- interim nodes: holds a hash value which is defined as hash of hash value of left and right children.
+
+![node types image](/ledger/docs/node_types.png)
+
+A *path* is a unique address of a node storing a payload. Paths are derived from the content of payloads (see common/pathfinder).
+
+![paths image](/ledger/docs/paths.png?raw=true "paths")
+
+#### Operations
+
+**Get**: Fetching a payload from the binary Merkle tree is by traversing the tree based on path bits. (0: left branch, 1: right branch)
+
+**Update**: Updates to the tree starts with traversing the tree to the leaf node, updating payload, hash value of that node and hash value of all the ancestor nodes (nodes on higher level connected to this node).
+
+![update image](/ledger/docs/tree_update.gif?raw=true "update")
+
+**Prove**: A binary Merkle tree can provide an inclusion proof for any given payload. A Merkle proof in this context includes all the information needed to walk through a tree branch from an specific leaf node (key) up to the root of the tree (yellow node hash values are needed for inclusion proof for the green node).
+
+![proof image](/ledger/docs/proof.png?raw=true "proof")
+
+### binary Merkle trie
+A **binary Merkle trie** in this context is defined as a compact version of binary Merkle tree, providing exact same functionality but doesn't explicitly store empty nodes. Formally, a node is empty: 
+* the node is an empty leaf node (see definition above)
+* an interim node is defined to be empty, if its two children are empty. 
+
+![binary partial trie image](/ledger/docs/trie_update.gif?raw=true "binary partial trie")
+
+### forest 
+Formally, a **forest** is any acyclic graph. Any set of disjoint trees forms a forest. In the context of Flow, we take an existing state, represented by a Merkle tree. Updating the payload of some of the leafs creates a new Merkle tree, which we add to the forest. In other words, the forest holds a set of state snapshots
+
+### compact forest 
+A **compact forest** constructs a new trie after each update (copy on change) and reuses unchanged sub-tries from the parent.
+
+![compact forest image](/ledger/docs/reuse_sub_trees.gif?raw=true "compact forest")
+
+### path finder 
+**Path finder** deterministically computes a path for a given payload. Path finder is responsible to make sure the trie grows in balance and relevant data are stored nearby so that the proof size can be optimized.
```

Comment:
I think the "trie grows in balance" point and "data stored nearby" go in opposite directions : if we prioritize a balanced tree then we loose storage locality, and if we prioritize data locality we will end up with an unbalanced tree.
 
I suggest we remove the storage locality point since the implemented version of path finder is just guaranteeing a balanced tree through hashing (at least for now).  

```suggestion
**Path finder** deterministically computes a path for a given payload. Path finder is responsible for making sure the trie grows in balance.
```

Reply:
Applied at 2ba9d6c